// @quasar/quasar-app-extension-testing-e2e-cypress/cct-dev-server/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.injectQuasarDevServerConfig = void 0;
/* eslint-disable */
const path_1 = require("path");
function getPackageJson() {
    return require((0, path_1.join)(process.cwd(), 'package.json'));
}
async function getSharedConfigImports(quasarAppPackage, bundler) {
    var _a, _b, _c;
    const { default: extensionRunner } = await (_a = `${quasarAppPackage}/lib/app-extension/extensions-runner`, Promise.resolve().then(() => __importStar(require(_a))));
    const { default: getQuasarCtx } = await (_b = `${quasarAppPackage}/lib/helpers/get-quasar-ctx`, Promise.resolve().then(() => __importStar(require(_b))));
    const { default: QuasarConfFile } = await (_c = `${quasarAppPackage}/lib/quasar-${bundler === 'vite' ? 'config' : 'conf'}-file`, Promise.resolve().then(() => __importStar(require(_c))));
    return {
        extensionRunner,
        getQuasarCtx,
        QuasarConfFile,
    };
}
async function quasarSharedConfig(quasarAppPackage, bundler) {
    const { extensionRunner, getQuasarCtx, QuasarConfFile } = await getSharedConfigImports(quasarAppPackage, bundler);
    const ctx = getQuasarCtx({
        mode: 'spa',
        target: void 0,
        debug: false,
        dev: true,
        prod: false,
    });
    // register app extensions
    await extensionRunner.registerExtensions(ctx);
    return {
        quasarAppPackage,
        QuasarConfFile,
        ctx,
    };
}
async function quasarWebpackConfig(quasarAppPackage) {
    var _a;
    const { QuasarConfFile, ctx } = await quasarSharedConfig(quasarAppPackage, 'webpack');
    const { default: { splitWebpackConfig }, } = await (_a = `${quasarAppPackage}/lib/webpack/symbols`, Promise.resolve().then(() => __importStar(require(_a))));
    const quasarConfFile = new QuasarConfFile(ctx);
    try {
        await quasarConfFile.prepare();
    }
    catch (e) {
        console.log(e);
        return;
    }
    await quasarConfFile.compile();
    const configEntries = splitWebpackConfig(quasarConfFile.webpackConf, 'spa');
    return configEntries[0].webpack;
}
async function getQuasarViteSpaConfig(quasarAppPackage) {
    var _a;
    const { default: quasarSpaConfig } = await (_a = `${quasarAppPackage}/lib/modes/spa/spa-config`, Promise.resolve().then(() => __importStar(require(_a))));
    return quasarSpaConfig;
}
async function quasarViteConfig(quasarAppPackage) {
    const { QuasarConfFile, ctx } = await quasarSharedConfig(quasarAppPackage, 'vite');
    const quasarConfFile = new QuasarConfFile({ ctx });
    const quasarConf = await quasarConfFile.read();
    if (quasarConf.error !== void 0) {
        console.log(quasarConf.error);
    }
    const quasarSpaConfig = await getQuasarViteSpaConfig(quasarAppPackage);
    // [1] -> https://github.com/cypress-io/cypress/issues/22505#issuecomment-1277855100
    // [1] Make sure to use the root for predictability
    quasarConf.publicPath = '/';
    const result = await quasarSpaConfig['vite'](quasarConf);
    // [1] Delete base so it can correctly be set by Cypress
    delete result.base;
    return result;
}
function injectQuasarDevServerConfig() {
    const { devDependencies } = getPackageJson();
    const bundler = devDependencies.hasOwnProperty('@quasar/app-vite')
        ? 'vite'
        : 'webpack';
    const quasarAppPackage = bundler === 'webpack' &&
        !devDependencies.hasOwnProperty('@quasar/app-webpack')
        ? '@quasar/app'
        : `@quasar/app-${bundler}`;
    const configExtractor = bundler === 'vite' ? quasarViteConfig : quasarWebpackConfig;
    return {
        framework: 'vue',
        bundler,
        [`${bundler}Config`]: async () => {
            // Trying q/app-vite 2+ & q/app-webpack 4+ specs
            try {
                // we workaround the build system because
                // we explicitly need an "import()" statement
                // otherwise this will fail for q/app-vite (as the imported file is an ES6 module)
                const promise = eval('import(`${quasarAppPackage}/lib/testing.js`);');
                const { getTestingConfig } = await promise;
                return getTestingConfig();
            }
            catch (_) { }
            return configExtractor(quasarAppPackage);
        },
    };
}
exports.injectQuasarDevServerConfig = injectQuasarDevServerConfig;
